### PR TITLE
grpc: implement TTL manager to re-buffer pending messages for bidirectional gRPC

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <rei@tetrate.io>

Commit Message: grpc: implement TTL manager to re-buffer pending messages for bidirectional gRPC 
Additional Description: Related with #17486. This PR provides the TTL manager for buffered bidirectional gRPC messages.
Risk Level: Low
Testing: Unit
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by Shikugawa** [comment]
@yanavlasov PTAL

File: source/common/grpc/buffered_message_ttl_manager.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      timer_->enableTimer(
+          std::chrono::duration_cast<std::chrono::milliseconds>(earliest_timepoint - now));
+    }
+  }
+
+  Event::Dispatcher& dispatcher_;
+  std::chrono::milliseconds message_ack_timeout_;
+  BufferedAsyncClientCallbacks& callbacks_;
+  Event::TimerPtr timer_;
+  std::map<MonotonicTime, absl::flat_hash_set<uint64_t>, std::greater<>> deadline_;
```
Comment (lizan) [inline source/common/grpc/buffered_message_ttl_manager.h:60]: a map makes this code unnecessarily complex here, since MonotonicTime is keep increasing and message_ack_timeout doesn't change over the life of this class, you should be able to manage this in a list (deque). That removes the need of lower_bound search and you should just iterate from beginning to eliminate all expired messages.

Comment (Shikugawa) [reply source/common/grpc/buffered_message_ttl_manager.h:60]: We will run into the problems in the following case with deque, I think.

`t_i`: post `(1, 2, 3)` to expire to `t_{i+2}`. Now the next timer will be set to start at `t_{i+2}`.
`t_{i+1}` (`t_i` <= `t_{i+1}` < `t_{i+2}`): post `(4, 5)` to expire to `t_{i+3}.
`t_{i+2}`: delete only `(1, 2, 3)`.
`t_{i+3}`: delete only `(4, 5)`.

I used map here because deleting `(1, 2, 3)` in `t_{i+2}` requires O(N) search in case of deque. I forgot about it.

Comment (Shikugawa) [reply source/common/grpc/buffered_message_ttl_manager.h:60]: Oh, you can just use deque to take out something smaller than the current time. My mistake.

File: source/common/grpc/buffered_message_ttl_manager.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      timer_->enableTimer(message_ack_timeout_);
+    }
+  }
+
+  const std::queue<std::pair<MonotonicTime, absl::flat_hash_set<uint64_t>>>& deadline() {
+    return deadline_;
+  }
+
+private:
+  void checkMessages() {
```
Comment (yanavlasov) [inline source/common/grpc/buffered_message_ttl_manager.h:35]: nit: make the method name more descriptive, i.e. checkExpiredMessages() or some such.

**Review by yanavlasov — COMMENTED** [review]
Is this PR complete? I do not see how the `BufferedMessageTtlManager` is being used. As such it is hard to reason about some of the decisions, like using flat_has_set of message ids in the TTL manager.

/wait-any

**Comment by Shikugawa** [comment]
@yanavlasov Could you take a look?

**Comment by phlax** [comment]
ping @yanavlasov

**Comment by alyssawilk** [comment]
@wrowe ping!

File: source/common/grpc/buffered_message_ttl_manager.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+class BufferedMessageTtlManager {
+public:
+  BufferedMessageTtlManager(Event::Dispatcher& dispatcher, BufferedAsyncClientCallbacks& callbacks,
+                            std::chrono::milliseconds message_ack_timeout)
+      : dispatcher_(dispatcher), message_ack_timeout_(message_ack_timeout), callbacks_(callbacks),
+        timer_(dispatcher_.createTimer([this] { checkExpiredMessages(); })) {}
+
+  ~BufferedMessageTtlManager() { timer_->disableTimer(); }
+
+  void setDeadline(absl::flat_hash_set<uint64_t>&& ids) {
```
Comment (snowp) [inline source/common/grpc/buffered_message_ttl_manager.h:21]: maybe this should be `addDeadlineEntry` or something like that? `setX` sounds like we're overriding, but this is adding to the collection of deadlines

File: source/common/grpc/buffered_message_ttl_manager.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+#include <queue>
+
+#include "envoy/event/dispatcher.h"
+
+#include "source/common/grpc/buffered_async_client.h"
+
+namespace Envoy {
+namespace Grpc {
+
+class BufferedMessageTtlManager {
```
Comment (snowp) [inline source/common/grpc/buffered_message_ttl_manager.h:12]: I think this class could benefit from some comments, it's not immediately clear how this works or what it tries to solve

File: source/common/grpc/buffered_message_ttl_manager.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    while (!deadline_.empty()) {
+      auto& it = deadline_.front();
+      if (it.first > now) {
+        break;
+      }
+      for (auto&& id : it.second) {
+        // If the retrieved message is a PendingFlush, it means that the message
+        // has timed out. A timeout is treated as an error, and the callback will
+        // re-buffer the message.
+        callbacks_.onError(id);
```
Comment (snowp) [inline source/common/grpc/buffered_message_ttl_manager.h:47]: This class just operates on sets of ids that are expiring, so this talk about PendingFlush seems a bit out of place?

Comment (snowp) [inline source/common/grpc/buffered_message_ttl_manager.h:47]: I wonder if instead of reusing `BufferedAsyncClientCallbacks` this should have its own simpler callback interface? I think that might be a bit clearer? WDYT?

File: source/common/grpc/buffered_message_ttl_manager.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  void setDeadline(absl::flat_hash_set<uint64_t>&& ids) {
+    const auto expires_at = dispatcher_.timeSource().monotonicTime() + message_ack_timeout_;
+    deadline_.emplace(expires_at, std::move(ids));
+
+    if (!timer_->enabled()) {
+      timer_->enableTimer(message_ack_timeout_);
+    }
+  }
+
+  const std::queue<std::pair<MonotonicTime, absl::flat_hash_set<uint64_t>>>& deadline() {
```
Comment (snowp) [inline source/common/grpc/buffered_message_ttl_manager.h:30]: What's the use case of exposing this? Test only? If so I'd suggest naming this `deadlineForTest` or something to express this.

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks for working on this! The logic seems right to me, just a few comments to make this easier for people to understand

File: source/common/grpc/buffered_message_ttl_manager.h (hunk 5 above)
Comment (Shikugawa) [reply source/common/grpc/buffered_message_ttl_manager.h:47]: Does it mean that `BufferedMessageTtlManager` exposes some callbacks and is accessed by `BufferedAsyncClient`?

File: source/common/grpc/buffered_message_ttl_manager.h (hunk 4 above)
Comment (Shikugawa) [reply source/common/grpc/buffered_message_ttl_manager.h:14]: Added

File: source/common/grpc/buffered_message_ttl_manager.h (hunk 5 above)
Comment (snowp) [reply source/common/grpc/buffered_message_ttl_manager.h]: I was thinking we could just have a single `std::function<void(uint64_t)> expiry_callback` function passed to the constructor, that way we don't have any real coupling between this and the buffered async client, this just becomes a "pure" TTL management class

Comment (Shikugawa) [reply source/common/grpc/buffered_message_ttl_manager.h]: fixed.

File: source/common/grpc/buffered_async_client.h [hunk 7]
```diff
@@ -12,39 +14,20 @@ namespace Grpc {
...
   BufferedAsyncClient(uint32_t max_buffer_bytes, const Protobuf::MethodDescriptor& service_method,
                       Grpc::AsyncStreamCallbacks<ResponseType>& callbacks,
-                      const Grpc::AsyncClient<RequestType, ResponseType>& client)
+                      const Grpc::AsyncClient<RequestType, ResponseType>& client,
+                      Event::Dispatcher& dispatcher, std::chrono::milliseconds message_timeout_msec)
       : max_buffer_bytes_(max_buffer_bytes), service_method_(service_method), callbacks_(callbacks),
-        client_(client) {}
+        client_(client),
+        ttl_manager_(
+            dispatcher, [this](uint64_t id) { this->onError(id); }, message_timeout_msec) {}
```
Comment (snowp) [inline source/common/grpc/buffered_async_client.h:30]: you don't need `this->`, just do `onError(id);`

File: test/common/grpc/buffered_async_client_test.cc [hunk 8]
```diff
@@ -23,117 +27,145 @@ namespace Envoy {
...
+  }
+
+  Api::ApiPtr api_;
+  Event::DispatcherPtr dispatcher_;
   const Protobuf::MethodDescriptor* method_descriptor_;
   envoy::config::core::v3::GrpcService config_;
   NiceMock<Upstream::MockClusterManager> cm_;
   NiceMock<Http::MockAsyncClient> http_client_;
+  Http::MockAsyncClientStream http_stream_;
+  DangerousDeprecatedTestTime test_time_;
```
Comment (snowp) [inline test/common/grpc/buffered_async_client_test.cc:94]: Can we use simulated test time here instead?

File: test/common/grpc/buffered_message_ttl_manager_test.cc [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+      : api_(Api::createApiForTest()), dispatcher_(api_->allocateDispatcher("test_thread")) {}
+
+  Api::ApiPtr api_;
+  Event::DispatcherPtr dispatcher_;
+  std::shared_ptr<BufferedMessageTtlManager> ttl_manager_;
+  std::chrono::milliseconds msec_{1000};
+  uint32_t callback_called_counter_ = 0;
+};
+
+TEST_F(BufferedMessageTtlManagerTest, Basic) {
```
Comment (snowp) [inline test/common/grpc/buffered_message_ttl_manager_test.cc:29]: This test is a bit hard to follow, maybe add some comments?

File: test/common/grpc/buffered_message_ttl_manager_test.cc [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        }
+        ++callback_called_counter_;
+      },
+      msec_);
+  ttl_manager_->addDeadlineEntry(std::move(ids));
+
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
+  EXPECT_EQ(ttl_manager_->deadlineForTest().size(), 0);
+
+  // Test if deadline queue is empty after queue cleared once.
```
Comment (snowp) [inline test/common/grpc/buffered_message_ttl_manager_test.cc:56]: Should this comment be before L54?

File: source/common/grpc/buffered_message_ttl_manager.h [hunk 11]
```diff
@@ -4,23 +4,25 @@
...
 
-#include "source/common/grpc/buffered_async_client.h"
-
 namespace Envoy {
 namespace Grpc {
 
+using BufferedMessageExpirationCallback = std::function<void(uint64_t)>;
+
 // This class is used to manage the lifetime of messages stored in BufferedAsyncClient. Messages
 // whose survival period has expired will be deleted from the Buffer.
```
Comment (snowp) [inline source/common/grpc/buffered_message_ttl_manager.h:13]: Could you add a bit to this description that talks about how this acts on a set of ids that are all expired at the same time? I think that might not be immediately obvious

Thanks!

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks for iterating on this! A few more comments

File: source/common/grpc/buffered_message_ttl_manager.h (hunk 11 above)
Comment (Shikugawa) [reply source/common/grpc/buffered_message_ttl_manager.h]: added

File: test/common/grpc/buffered_message_ttl_manager_test.cc (hunk 10 above)
Comment (Shikugawa) [reply test/common/grpc/buffered_message_ttl_manager_test.cc:57]: fixed

File: test/common/grpc/buffered_async_client_test.cc (hunk 8 above)
Comment (Shikugawa) [reply test/common/grpc/buffered_async_client_test.cc]: Simulated test time is no longer needed here. Thus I deleted it.

File: source/common/grpc/buffered_message_ttl_manager.h [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+// whose survival period has expired will be deleted from the Buffer.
+// We can set the ID you want to monitor TTL in the TTL Manager,
+// which will set up a callback to check for expiration.
+// You can set the ID even after the callback has been invoked.
+// When the callback is invoked, the callback given to the constructor will be
+// executed with the TTL-elapsed ID as an argument.
+// After that, if the ID to be monitored is not empty, the callback for expiration check will be set
+// again. The TTL Manager can be given a set of IDs that are expected to expire at the same time.
+// When checking for ID expiration, an expiration callback will be called for each ID
+// belonging to this set of IDs.
```
Comment (snowp) [inline source/common/grpc/buffered_message_ttl_manager.h:22]: I think it might be more natural to fold this into the start of this paragraph and talk about it handling sets of IDs + triggering callbacks for each expired ID at the beginning.

I'd also call out that there is no mechanism for preventing an expiration, so it's expected that the receiver of the callback to handle that case. 

Lastly, I'd have this comment also mention that we're using a fixed deadline per inserted entry, vs one that is configurable at insertion time

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks! A few comments on the documentation wording

File: source/common/grpc/buffered_message_ttl_manager.h [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+// The TTL Manager can be given a set of IDs that are expected to expire at the same time.
+// When checking for ID expiration, an expiration callback will be called for each ID belonging to
+// this set of IDs. This class is used to manage the lifetime of messages stored in
+// BufferedAsyncClient. Messages whose survival period has expired will be deleted from the Buffer.
+// It allows monitoring a set of IDs for expiration, triggering a callback upon expiration.
+// This class is able to monitor multiple sets of IDs at the same time, even after some of them have
+// expired. When the callback is invoked, the callback given to the constructor will be will be
+// executed for each of the TTL-elapsed IDs. After that, if the IDs to be monitored is not empty,
+// the manager will continue to monitor for expiration again.
```
Comment (snowp) [inline source/common/grpc/buffered_message_ttl_manager.h:20]: This still seems a bit repetitive, how about:

```
This class is used to manage the TTL for pending uploads within a BufferedAsyncClient. Multiple IDs
can be inserted into the TTL manager at once, all with the same TTL (specified in the constructor). Upon expiry,
the TTL manager will invoke the provided expiry callback for each ID. Note that there is no way to disable
the expiration, and so it's up to the recipient of the callback to handle this. BufferedAsyncClient will do
the right thing here: if the expired ID is still in flight it will be returned to the buffer, otherwise it does nothing.

The TTL manager is designed to handle multiple sets of IDs inserted at various times, backing this with a
single Timer. This allows us to track a large amount of IDs inserted at different times without using a lot of
different timers, which could put undue pressure on the event loop.
```

?

**Review by snowp — CHANGES_REQUESTED** [review]
A suggestion for the comment wording, otherwise LG

/wait

File: source/common/grpc/buffered_message_ttl_manager.h (hunk 13 above)
Comment (Shikugawa) [reply source/common/grpc/buffered_message_ttl_manager.h]: Thank you for your suggestion!
